src/Proto3/Wire/Decode.hs

```diff
@@ -441,6 +443,26 @@ sfixed64 = runGetFixed64 getInt64le
 at :: Parser RawField a -> FieldNumber -> Parser RawMessage a
 at parser fn = Parser $ runParser parser . fromMaybe mempty . M.lookup fn
 
+-- | Try to parse different field numbers with their respective parsers.
+-- This is used to express alternative between possible fields
+--
+-- If no field number are present for the oneof, then the default
+-- behavior of the first parser is used.
+--
+-- TODO: contrary to the protobuf spec, in the case of multiple fields number
+-- matching the oneof content, the choice of field is biased
+-- to the order of the list, instead of being biased to the last field
+-- of group of field number is the oneof. This is related to the Map
+-- used for input that preserve order across multiple invocation of the same
+-- field, but not across a group of field.
+oneof :: [(FieldNumber, Parser RawField a)] -> Parser RawMessage a
+oneof []                        = Parser $ const $
+    Left (BinaryError "invalid oneof with no field. oneof need to contains at least 1 element")
+oneof parsers@((_,defParser):_) = Parser $ \input ->
+    case msum $ map (\(fn, parser) -> (parser, ) <$> M.lookup fn input) parsers of
+        Nothing                -> runParser defParser mempty
```

Comment:
I know that this matches the `at` combinator, but I think the `at` combinator's behavior is not correct.  It was written with the expectation that all `Parser`s would fail on empty input, but we cannot guarantee that because:

* The `Parser` constructor is exposed, so users can create a custom parser that always succeeds
* `Parser` implements `Applicative`, so `pure` is a valid `Parser` that always succeeds on empty input

I think the right behavior here (and for `at`) is to fail with a descriptive `Left` if there is no matching field